### ISSUE TITLE
Theme sync does sync for whole folder content

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#10123]
2021.03.07-NB

Yesterday I was contacted by a geocaching friend, who also uses our nightly build.
He contacted me because on his device (Samsung S10, Android 9) the app settings suddenly claim, that c:geo eats up >2GB of internal storage additionally to some days before.

After remotely trying to analyze the problem it turned out he places his themes and maps into the same folder and (correctly) selected the same SAF folder for both map and themes. This obviously triggers c:geo not only to copy the themes into the internal sync folder but also all his maps.

I can imagine, that this could be a huge problem for many users out there. I do also know people, who do not use a very clean data structure on their devices. For example I had someone on support who uses his system download folder as maps and themes folder, because this is the place where they are stored after (manually...before we introduced the downloader) downloading them from the maps server. This download folder might (over time) contain unlimited more files totally unrelated to c:geo.

If my assumption is correct, that we simply sync all contents of the selected themes folder, I suggest we need a clever selection algorithm to distinguish the file. As there are several file types involved, we can probably only select files by their size (I guess the theme XML and PNGs are usually all <<1MB) to limit the amount of unneeded garbage we grab with our sync.

**Comment by Lineflyer** [comment]
Additional question:
While reproducing the problem on my Google Pixel3a, Android 11 c:geo started up normally after selecting the map folder as theme folder and the app manager showed the large size for the app. However on consecutive starts c:geo hangs on splash screen. Not sure if this is related.

**Comment by eddiemuc** [comment]
Ah, again something I did not think of....

Some thoughts:
* Simply limiting to files <1MB might work, but we have to be careful how to select the boundary because we now also support zipped themes, and they are a bit larger than just an XML. E.g. the famous Elevate.zip is about 440KB in size. Maybe a limit of 800KB would be approbiate?
* Another approach is to sync only certain file suffixes (xml, zip, png, jpg, maybe a few others), or do negative filtering (don't copy files with suffix map)
* OR we don't do the sync at all. @Lineflyer there's still the test code in the nightly allowing you to select the themens DIRECTLY from the public folder when in debug mode. Have you tried how performance is for those on a real Android 11 device?

**Comment by eddiemuc** [comment]
> Additional finding:
> While reproducing the problem on my Google Pixel3a, Android 11 c:geo started up normally after selecting the map folder as theme folder and the app manager showed the large size for the app. However on consecutive starts c:geo "hangs" on splash screen for >20 seconds. Not sure if this is related. Do we sync while splash screen is shown?

Sync should start as part of MainActivity.OnCreate and should be asynchronous. Don't know exactly when splash screen starts showing and when it stops from a code perspective.  Anything in the logs? (in debug mode, log entries should be visible on INFO level showing start and stop of theme folder synchronization)

**Comment by Lineflyer** [comment]
> Some thoughts:

> Simply limiting to files <1MB might work, but we have to be careful how to select the boundary because we now also support zipped themes, and they are a bit larger than just an XML. E.g. the famous Elevate.zip is about 440KB in size. Maybe a limit of 800KB would be approbiate?

> Another approach is to sync only certain file suffixes (xml, zip, png, jpg, maybe a few others), or do negative filtering (don't copy files with suffix map)

Is there any definition of valid file types which themes may consist of?
Still in a scenario with having a size limit of <1MB (or 800kB) and the needed file types, it might still copy a lot of stuff if I select my whole storage.

> OR we don't do the sync at all. @Lineflyer there's still the test code in the nightly allowing you to select the themens DIRECTLY from the public folder when in debug mode. Have you tried how performance is for those on a real Android 11 device?

That would mean we trash a lot of coding you have done :(  but it would indeed be the safe solution.
I checked it again and made a video:
- Starting with synced un-zipped theme
- Change to same theme on public folder
- Change to synced zipped theme
- Change to zipped theme in public folder

My personal conclusion:
As changing a theme is not an everyday task, the noticeable slower progress would be acceptable here.
BUT sadly the same delay as after theme change also applies each time when I start the map, which is not acceptable IMHO.

Idea:
Couldn't we sync only the selected theme as soon as the user selects it? Sounds simple for a zipped theme, but for an unzipped theme we would probably need to read the XML to identify the needed resource files (icons, etc.) and sync them. 
The huge advantage would be that there is no backup magic on each startup but only on user interaction and for sure only targeting a real theme.

**Comment by Lineflyer** [comment]
Just to be complete (no idea, whether this would be reasonable):
We could also zip unzipped themes in the public folder once they are selected...but that's maybe unexpected for the user.

**Comment by eddiemuc** [comment]
> Idea:
>Couldn't we sync only the selected theme as soon as the user selects it? Sounds simple for a zipped theme, but for an unzipped theme we would probably need to read the XML to identify the needed resource files (icons, etc.) and sync them.
The huge advantage would be that there is no backup magic on each startup but only on user interaction and for sure only targeting a real theme.

I see the reason behind this idea, but I really think it gets too complicated at this point. There must be a more straightforward way.

**Comment by Lineflyer** [comment]
> There must be a more straightforward way.

Drop the sync implementation and inform users of unzipped themes to use zipped themes to improve performance.
Such info needs to be done in a proper way within c:geo (having it only in FAQ will reach less users).

**Comment by Lineflyer** [comment]
Does the mapsforge library support any buffering of themes (like it does for tiles)?

**Comment by fm-sys** [comment]
> Drop the sync implementation and inform users of unzipped themes to use zipped themes to improve performance. 

That would IMHO be ok. Users using themes are kind of advanced anyway, so zipping shouldn't be difficult for them. 

Sorry @eddiemuc that parts of your great work might need to be removed, but I guess implementing the ziped theme provider for mapsforge was even more work, so this would at least come to bear... 

> Such info needs to be done in a proper way within c:geo (having it only in FAQ will reach less users). 

Time for another OneTimeMessage ;-) But please display it only once a user uses a different theme then the default one, so that we do not disturb users with information they don't care about...

**Comment by eddiemuc** [comment]
> Does the mapsforge library support any buffering of themes (like it does for tiles)?

Yes it does. Once a theme is read into memory, it is caches there. This is dropped again on map recreation of course (which we do when you leave the map viewer and then open it again)

**Comment by eddiemuc** [comment]
> > There must be a more straightforward way.
> 
> Drop the sync implementation and inform users of unzipped themes to use zipped themes to improve performance.
> Such info needs to be done in a proper way within c:geo (having it only in FAQ will reach less users).

Valid approach, but I have to ask specifically that performance is tested for all supported Android versions before doing that. I found huge differences in file access performance across versions, funnily enough newer/higher versions had worse performance (Android 11/SDK30 was the worst), and at least in emulator SDK30 performance was pretty bad even when using a ZIP

**Comment by eddiemuc** [comment]
> Sorry @eddiemuc that parts of your great work might need to be removed, but I guess implementing the ziped theme provider for mapsforge was even more work, so this would at least come to bear... 

No worries here :-)
I would, for now, only remove the function access to it, not the sync code itself, at least not until after the beta phase. So in case sat happens we can go back to it.

> > Such info needs to be done in a proper way within c:geo (having it only in FAQ will reach less users). 

This also means that Multitheme approaches as done by very experienced expert users like @Lineflyer and @magpiefourtytwo will unfortunately not work (with acceptable performance) any more

**Comment by MagpieFourtyTwo** [comment]
> This also means that Multitheme approaches as done by very experienced expert users like @Lineflyer and @MagpieFourtyTwo will unfortunately not work (with acceptable performance) any more

This would be a *big* regression and I would highly appreciate, if we could keep at least the Multitheme functionality in any way.
While I would rate zip support as pure convenience feature, the themes themselves are not "just" additional function, but for scenarios like with @Lineflyer's setup, it's a major performance issue.

**Comment by Lineflyer** [comment]
> This also means that Multitheme approaches as done by very experienced expert users like @Lineflyer and @MagpieFourtyTwo will unfortunately not work (with acceptable performance) any more

It doesnt matter how many themes you have. Without sync my usecase does even improve in performance...as not everything in the themes folder is synced...or did I miss something here?

**Comment by MagpieFourtyTwo** [comment]
> Without sync my usecase does even improve in performance

I'm confused. I thought, sync has been invented as a workaround for deadly slow SAF access to the themes folder(s) ...
But as SAF will come for sure, AFAICS, how can it now be faster for you without any sync at all ...?

**Comment by eddiemuc** [comment]
I might have introduced some misunderstanding before, so let me try again:

* If we remove the folder sync thing, then of course all theme related files are read directly from the public folder using SAF
* Both the "extracted" way (xml files with subdirs containing icons etc) and the Zip-way will still work
* **BUT** due to slowness of SAF, the "extracted" way will be very very slow - probably so slow that users might think it is broken (to open the normal Elevate.xml theme with a few hundert icons, it took me about 10 seconds on Android 11 emulator. This wll happen every time you open the map viewer or change the theme). The Zip-way will also be slow, but since only one file is read and its content is cached across map viewer accesses (within the same c:geo session) this might not be as significant.

**Comment by eddiemuc** [comment]
Adding to the previous comment: it would be possible to write a "cache" object to also cache the icons in memory read from "extracted" themes (although it woould be some work) and to keep this in memory across multiple openings of map viewer. But this cache must be invalidated on map theme change (-> 10 seconds wait), and of course it needs to be filled on first map opening after restart of c:geo (-> 10 seconds wait)

**Comment by MagpieFourtyTwo** [comment]
> so let me try again

Yep, that's the way I have understood it.
(although I think "have been understanding" would have been a more correct translation - but am I the only one who thinks this is extremely ugly and bulky?)

> it would be possible to write a "cache" object ...

Sorry, no way. Waiting more than a second is not really tolerated by most GUI users.
And things get even worse the more aggressive Android gets WRT removing stuff from memory.

**Comment by eddiemuc** [comment]
> > so let me try again
> 
> Yep, that's the way I have understood it.
> (although I think "have been understanding" would have been a more correct translation - but am I the only one who thinks this is extremely ugly and bulky?)

:-)

> > it would be possible to write a "cache" object ...
> 
> Sorry, no way. Waiting more than a second is not really tolerated by most GUI users.
> And things get even worse the more aggressive Android gets WRT removing stuff from memory.

...also I spoke too soon. It would indeed not be possible to write such a cache object. Because I would face the same challenge as with the sync: deciding what to put in and what not. Worst case I end up pulling an offline map into memory. Not a good idea...

**Comment by Lineflyer** [comment]
> > Without sync my usecase does even improve in performance
>
> I'm confused. I thought, sync has been invented as a workaround for deadly slow SAF access to the themes folder(s) ...
But as SAF will come for sure, AFAICS, how can it now be faster for you without any sync at all ...?

What I meant is:
If we drop the sync, it does not matter, whether you have only one or tons of themes in your theme folder as there is no need to copy all of them initially and check all of them periodically (even while its happening in background I do still see low performance of the app while it is in progress).
Of course (as mentioned above) the downside is the noticeable delay on switching to another theme (not so relevant) and on each start of the map (most relevant).

@eddiemuc 
On my device (as can be seen in the video) its not 10 seconds to load a theme, more about 3-5 seconds. Still bad.
But as you said regarding "Use a straightforward solution":
Caching, file filtering, file extension guessing or even XML analyzing are all complex solutions, which might - after all - fail under some circumstances (or if the user does "dumb" things).
So also after sleeping a night over it, it pretty much seems its IMHO the best solution to rely on SAF only....but how to instruct the user, that he should change to zipped themes?

**Comment by eddiemuc** [comment]
I can provide a PR to kill the sync as soon as time permits. For the rest - I don't really have a good idea. If people use the new map theme download option, then they get zips anyway.

**Comment by Lineflyer** [comment]
@cgeo/team Other ideas?

Regarding user notification:
Can we display a one-time-message under the condition, that the user has a theme (other than default) in use?

**Comment by fm-sys** [comment]
> Regarding user notification:
Can we display a one-time-message under the condition, that the user has a theme (other than default) in use?

That's the easiest part ;-) I can provide a PR if wanted...

**Comment by moving-bits** [comment]
Hm, difficult situation. How about this:

- Leave sync in, but disable it on default.
- Show a one-time message if the user uses a theme other than default. This message could explain how to use zipped themes or how to activate syncing.
- Syncing can be enabled using settings. There should be an explanation that you should use a dedicated folder for themes.

**Comment by fm-sys** [comment]
> Hm, difficult situation. How about this:
> 
>     * Leave sync in, but disable it on default.
> 
>     * Show a one-time message if the user uses a theme other than default. This message could explain how to use zipped themes or how to activate syncing.
> 
>     * Syncing can be enabled using settings. There should be an explanation that you should use a dedicated folder for themes.

Sounds good. But I would explicitly annotate that setting as experimental ;-)

**Comment by MagpieFourtyTwo** [comment]
Talked to a guy today who was completely convinced it was an extremely smart idea of him to configure his maps and themes folder to ``/storage/0/Downloads``. "This way I don't have to additionally move all the stuff after downloading ..."

Horrific, WRT sync, of course, but if you think about it for a moment, one might really think that's pretty reasonable. So I wouldn't be surprised if there are a few more out there in the wild who think they're just as smart.

Hence it would probably be good to take the user by his hand and try to avoid at least the biggest problem potentials. Like a "it's highly recommended to use a dedicated folder" message when he tries to configure ``*/Download`` as map or themes folder. Or if he tries to configure the themes folder to a folder, where already ``.map`` files (resp. *big* ones [>10 MB?]), are living. Or if he tries to configure themes folder and map folder to the very same target ...?

Or ... c:geo may even become a little bit more restrictive and automatically sets (or at least defaults) the themes folder to ``%MapFolder%/_themes`` ...?

**Comment by eddiemuc** [comment]
> Or ... c:geo may even become a little bit more restrictive and automatically sets (or at least defaults) the themes folder to %MapFolder%/_themes ...?

Actually, there is already a default value for the folder, it is %baseFolder%/themes (or something like this). But upon migration, of course c:geo takes over configured legacy values.

Taken into consideration all that was said, I offer now to do the following:
* Take out the sync by default
* Make a new setting under map themes folder to turn it on or off
* Adjust explanatory text for map theme folder (there is already one in the settings for it, it also points to FAQs)
* Implement an absolute maximum sync file size as well as an "exclude" file filter (will not sync Files with suffix map)
* Show a warning dialog when user selects "sync on" bit had more than x MB of data in the current theme folder

What do you think?

**Comment by MagpieFourtyTwo** [comment]
> %baseFolder%/themes

Yep, know it, and don't like it, as IMHO it's an a) completely superfluous setting, just making it harder for Jo(s)e(phine) Average to configure offline maps properly, although it's a setup step which can easily be omitted, as other programs happily do, like Locus which simply expects ``_themes`` to be within the map folder, b) suggests that themes should be exclusively stored (thus copied) for c:geo and should not be shared with other apps, is c) even more superfluous, if you have Locus installed anyway, with ``%MapFolder%`` already configured to ``/storage/<whatever>/Locus/mapsVector`` and d) on every start creates a folder ``%baseFolder%/themes``, polluting your base folder (while this may have been cured in the meantime, can't find it right now, but this may be different again on next re-install) and last but not least it most obviously introduces massive sync-problems.  ;)

Thus I would suggest to just drop it. Would solve a bunch of problems in just one go.

> Make a new setting under map themes folder to turn it on or off

JTBC (you most probably have it in mind, anyway): Disabling the sync option should immediately clean the internal sync folder, too.
Otherwise it would be kind of "hard" (read "not obvious") to get rid of erroneous sync'ed stuff.

> What do you think?

I'm pretty sure some ingenious DAU will still find another way to cheat sync to death, but for the time being it sounds really good.  As usual. ;)

**Comment by eddiemuc** [comment]
> JTBC (you most probably have it in mind, anyway): Disabling the sync option should immediately clean the internal sync folder, too.
Otherwise it would be kind of "hard" (read "not obvious") to get rid of erroneous sync'ed stuff.

Yes, had that in mind :)

**Comment by eddiemuc** [comment]
> > > %baseFolder%/themes
> > 
> > Yep, know it, and don't like it 
> [...] 
> > Thus I would suggest to just drop it. Would solve a bunch of problems in just one go.
> 
> **Please don't.** I (and probably many other users) have all folders at their default location under the base directory, which means that I/he only need to grant folder permission for the base dir and currently every thing else works "out of the box". This is a) way more easy for users who don't use themes at all or don't care about the folder location and b) is much faster for me when switching between nightly and developer builds

I understand that the suggestion is to make the default theme folder `%OfflineMapsFolder%/_themes` instead of `%baseFolder%/themes`. That's a valid request, and still users using default setting setup only have one permission to give (because OfflineMapsFolder is subfolder of base as default). Changing this would be trivial. If this is wanted I can do it easily?

**Comment by baiti** [comment]
Recently, while debugging some of my Adventure Lab stuff I stumbled over some strange stuff in the wizard. It was related to offline maps and "sync" appeared in the UI somewhere. That was the point I backed out of the wizard for good.

Here is how I dealt with offline maps so far and I hope it will continue to be that way in the future: I use Locaus Map and cgeo and I try my best to make them both use the same instances of offline maps (and themes for that sake). And so far that all worked nicely.

I download offline maps myself and I see no point in seekting Locus' nor cgeo's help with that process. I put those maps into a folder /OfflineMaps, also the Theemes, if there are any. 

In Locus' settings I point offline maps to -> /OfflineMaps
In cgeo's settings I point offline maps to -> /OfflineMaps

and that was it ... for years.

Is this going to change with the next release of cgeo? I realize some aspects perhaps have to change because of scoped storage and / or SAF? I don't know what a "sync" is in that context and I hope I have a choice to not use any of this?

Is the new design documented somewhere? I believe it would be very important for users to really and down to the wire understand what changes and why.

If this is just me having no clue what's going on or being ignorant, please ignore this post. If you want me to test some of the newer work and provide feedback afterwards, let me know and I will do.

**Comment by baiti** [comment]
Wow! I begin to understand. My first reaction: This will lead to major desaster starting November this year. I would anticipate that there are **many** applications out there which will either no longer work or no longer being available on Play Store or not being allowed to push updates into the Store because certain requirements aren't met.

I can understand the aspects of access permissions but what you say under "sync", that makes me worry a lot. It seems like what cgeo is about to do is a "circumvention for a problem" (as opposed to a solution for a problem)  introduced by "G". I wouldn't be surprised if they just  **have** to take that back.

**Comment by baiti** [comment]
I'd assume the "sync" provides some sort of caching in order to greatly speed up file access? So this might be the only viable solution. Could it be done such that the user doesn't even have to know? I have to admit, when I walked the wizard last time, I didn't pay too much attention to those aspects due to my lack of understanding the motivation for it and the reasoning.

**Comment by baiti** [comment]
I'd assume, sync means you copy the data to some place where you have higher speed access? In this case this might mean a twice as heavy footprint on storage, which in turn may be less and less of a problem as the storage sizes in these phones grow dramatically. I still use a 64GB device w/o SD card (for performance). Would the sync approach even work in a scenario where my 64GB total device has only 10GB left and I have, let's say 8GB of offline maps on the device?

**Comment by eddiemuc** [comment]
@baiti yes, the sync was meant as some sort of cache exactly like you said. It was planned to be absolutely transparent to the user.

But during the way we discovered some practical problems, e.g. what if the user places his theme files in download folder or some device root folder and sync starts copying very large directory structures into app-internal storage. This is why we are currently take a step back.

**Comment by baiti** [comment]
@eddiemuc .. now the loop closes. I started to follow the thread where @MagpieFourtyTwo elaborated on that scenario.

Just on the side: I am using the "Fieldnotes" plugin of Locus Map and in the process of debugging my "kotline pusher app" (you know what I mean) I found out that this plugin dies on Android 11 when I press the "Share" button. They cannot even get to the point where my intent receiver gets called. I assume this is because they try to store a file into a place whether the SAF / scoped storage doesn't allow them to write to. I have contacted the developers 2 months ago ... no response. I'd assume, programs like this one will just cease to work. Worries ahead ....

**Comment by eddiemuc** [comment]
@baiti I find it hard to predict what will really happen with this whole scoped storage topic. If by all means it turns out that many of the major apps won't provide scoped storage updates until November, who knows what happens. Google might even shift the deadline for app compliance.

**Comment by baiti** [comment]
So back to the sync approach. Look at the scenario I described earlier. Offline maps are in "/OfflineMaps". I point cgeo to that path and Locus Map on that path. cgeo eventually does the sync thing for performance. I'd assuem Locus comes to similar ideas and does some sync approach too. Wouldn't that inevitably lead to having 3 copies of the data?

**Comment by eddiemuc** [comment]
I will not speculate how Locus or other apps will approach this topic :) 
The sync (which is in the end a copy of the data) only makes sense if the data amount is relatively small. A normal theme should take about 500KB of space. I always assumed that theme data should not take more than 2-3MB for average users.

However, as said before, I will remove sync (by default) anyway

**Comment by baiti** [comment]
> @baiti Only briefly for clarification: Syncing is done only for themes, not for map files. Large file sizes aren't a problem with SAF, only many very small files...

Yes, I just looked at what's in the /OfflineMaps  in my case. But isn't there an impiled assumption on where the Themes are relative to where the Maps are? For instance, in Locus, I can point Locus to a single directory only and it finds Maps and Themes   relative to that base path. If I wanted to move/copy the Themes, I may have to move the Maps too.

Does Android support something like symbolic links? that could help in this case.

**Comment by MagpieFourtyTwo** [comment]
> But isn't there an impiled assumption on where the Themes are relative to where the Maps are? For instance, in Locus, I can point Locus to a single directory only and it finds Maps and Themes relative to that base path. If I wanted to move/copy the Themes, I may have to move the Maps too.

Thanks for making [my point](https://github.com/cgeo/cgeo/issues/10123#issuecomment-795284840). ;)

> Does Android support something like symbolic links? that could help in this case.

Android itself does indeed support links, but first of all the currently used file system must be able to handle them.  Most regretfully the type of fs used is brand/manufacturer dependent, thus sym and hard llinks are not guaranteed to work in all cases.
If your fs in question is e. g. ext[234] or YAFFS2 (Androids current default for system partition, AFAIK), it will most probably work, but with e. g. FAT16, FAT32 or even exFAT (Android's defaults for external storage, which can even be read/written by M$) it will not.
And although you *might* succeed in creating a sym link from ext[234] to \<any>FAT, it will not work vice versa,

But why using links at all for map and themes data? Cause you can configure these paths anyway ...

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo , @baiti your posts above were for me the first time I ever heard that there is a default themes  folder under maps folder in c:geo. As far as I remember, c:geo always handled maps and themes as two different folders to be configured (or I missed something in SAF migration), and if no theme folder was configured then no themes were used.

It would be trivial (one-line-change) to change the default folder from `%base%/themes` to `%offlinemapfolder%/_themes` As far as I can think it would also not have any side effects. Since the default folder exists so far only in nightlies, there is also no migration necessary. So should I do this?

**Comment by eddiemuc** [comment]
> @MagpieFourtyTwo , @baiti your posts above were for me the first time I ever heard that there is a default themes  folder under maps folder in Locus. As far as I remember, c:geo always handled maps and themes as two different folders to be configured (or I missed something in SAF migration), and if no theme folder was configured then no themes were used.
> 
> It would be trivial (one-line-change) to change the default folder from `%base%/themes` to `%offlinemapfolder%/_themes` As far as I can think it would also not have any side effects. Since the default folder exists so far only in nightlies, there is also no migration necessary. So should I do this?

**Comment by baiti** [comment]
> But why using links at all for map and themes data? Cause you can configure these paths anyway ...

Links popped up in the context of "sync".  Since themes are in a subdirectory relative to where the maps are and I want to sync themes to another place but not maps and I don't want to break the association I thought why don't we sync the base directory and put in links for the map data to the original place. So the big files stay in place and the links maintain semantic association with the themes.

However, with what you elaborated on the filesystem's capabilities WRT symlinks that doesn't seem to be a good idea anyway.

**Comment by baiti** [comment]
> @MagpieFourtyTwo , @baiti your posts above were for me the first time I ever heard that there is a default themes folder under maps folder in **Locus**. As far as I remember, c:geo always handled maps and themes as two different folders to be configured (or I missed something in SAF migration), and if no theme folder was configured then no themes were used.
> 
> It would be trivial (one-line-change) to change the default folder from `%base%/themes` to `%offlinemapfolder%/_themes` As far as I can think it would also not have any side effects. Since the default folder exists so far only in nightlies, there is also no migration necessary. So should I do this?

I just inspected one of my devices. In this setup I started with Locus long ago and the offline maps  are in 
`/storage/emulated/0/Locus/mapsVector`, the themes in `/storage/emulated/0/Locus/mapsVector/_themes`.

In the days before the cgeo wizard I had configurued cgeo's path to the offline maps via
`Settings -> Map -> Directory with offline maps ======>  /storage/emulated/0/Locus/mapsVector`
`Settings -> Map -> Map themes directory ======>  /storage/emulated/0/Locus/mapsVector/_themes`

Now we can make some observations:

1. Locus seems to imply that themes are in a subdirectory of the directory where the maps are.
2. cgeo seems to be more flexible as it supports referencing the maps and themes in freely configurable directories, themes do not have to be in a subdirectory of where the maps are
3. If we want to configure a setup that allows using a single copy of the maps for both, Locus and cgeo, we'd have to adapt to the more restrictive apps's requirements, which is Locus. In other words we can't use cgeo's flexibility to place the two entities in arbitrarily chosen directories.
4. Based on what I believe I have learned now, `/storage/emulated/0/Locus` is an application's (Locus) private directory and starting November this year there will no longer even be an opt in to grant access to other applications (cgeo) to access this directory and everything beneath.

With those points made, I would assume that I need to put maps and themes outside of any application's private directories into a place where Android 11 facilitates access to after requesting permissions from the user and if the user decides to grant that access.

I further assume that the system has established some rules on the directory paths which support grant based access for any application that asks for permission. Let's assuem that is `/storage/emulated/0/Public`. What happens if I'd put the maps and themes into `/storage/emulated/0/Public/OfflineMaps` and I grant access to Locus and cgeo, would I then run into the scenario that @MagpieFourtyTwo described, that access to many many small files would take forever?

I would assume that a directory path with those properties (accessible, after request / grant) to multiple applications, they (G) don't do all the nasty stuff that slows it down so much. If that is true, the whole sync thing wouldn't be necessary, presumably. Please correct me if I'm wrong.

If point 4 above is also true, starting in November, in order to share a single copy of the maps a user **must** move those maps out of private app storage to some public place otherwise it will not work, with or without caching (sync). Is that correct?

**Comment by Lineflyer** [comment]
So much (off topic) activity here, it really hard to follow if topics mix up. Why don't you just open dedicated issues for dedicated topics? 

@eddiemuc 
Regarding the original topic:

>    Take out the sync by default

OK for me.

>    Make a new setting under map themes folder to turn it on or off

Yes, but default off.

>    Adjust explanatory text for map theme folder (there is already one in the settings for it, it also points to FAQs)

Good idea.
  
>  Implement an absolute maximum sync file size as well as an "exclude" file filter (will not sync Files with suffix map)

Yes, could be a good additional safety measure.

> Show a warning dialog when user selects "sync on" bit had more than x MB of data in the current theme folder

Yes, thats another safety measure 


Regarding the default location of the themes folder:
Indeed I can confirm, that Locus expects themes to be located in a subfolder `_themes` below the folder containing the vector offline maps. This is currently not configurable. So it would be wise to make this our default as well, so that users who download vector maps and themes with the c:geo downloader can easily reuse the structure with Locus.

Let me transfer this into a dedicated issue then.

**Comment by baiti** [comment]
> So much (off topic) activity here, it really hard to follow if topics mix up. Why don't you just open dedicated issues for dedicated topics?
> 

May I assume this is targeted at my latest post? May I further assume that what I posted is either (a) a non-issue, e.g. everybody except me knows how this works or (b) belongs to a different issue or thread?

If it is (a) I'll shut up and be quiet. If it is (b), what's the right place?

**Comment by Lineflyer** [comment]
Hi @baiti . No worries, I just created a new issue to discuss and clarify a good folder structure. I guess that was also the main topic of your post. Otherwise feel free to open another new issue.

I am a friend of having dedicated issues for each bug/improvement, so that we can keep the overview and track what is to be done and what is fixed already. Otherwise we are likely to end up with forgetting something or forever-open issues (based on my experience over years).

**Comment by baiti** [comment]
> > @MagpieFourtyTwo , @baiti your posts above were for me the first time I ever heard that there is a default themes folder under maps folder in **Locus**. As far as I remember, c:geo always handled maps and themes as two different folders to be configured (or I missed something in SAF migration), and if no theme folder was configured then no themes were used.
> > It would be trivial (one-line-change) to change the default folder from `%base%/themes` to `%offlinemapfolder%/_themes` As far as I can think it would also not have any side effects. Since the default folder exists so far only in nightlies, there is also no migration necessary. So should I do this?
> 
> I just inspected one of my devices. In this setup I started with Locus long ago and the offline maps are in
> `/storage/emulated/0/Locus/mapsVector`, the themes in `/storage/emulated/0/Locus/mapsVector/_themes`.
> 
> In the days before the cgeo wizard I had configurued cgeo's path to the offline maps via
> `Settings -> Map -> Directory with offline maps ======> /storage/emulated/0/Locus/mapsVector`
> `Settings -> Map -> Map themes directory ======> /storage/emulated/0/Locus/mapsVector/_themes`
> 
> Now we can make some observations:
> 
> 1. Locus seems to imply that themes are in a subdirectory of the directory where the maps are.
> 2. cgeo seems to be more flexible as it supports referencing the maps and themes in freely configurable directories, themes do not have to be in a subdirectory of where the maps are
> 3. If we want to configure a setup that allows using a single copy of the maps for both, Locus and cgeo, we'd have to adapt to the more restrictive apps's requirements, which is Locus. In other words we can't use cgeo's flexibility to place the two entities in arbitrarily chosen directories.
> 4. Based on what I believe I have learned now, `/storage/emulated/0/Locus` is an application's (Locus) private directory and starting November this year there will no longer even be an opt in to grant access to other applications (cgeo) to access this directory and everything beneath.
> 
> With those points made, I would assume that I need to put maps and themes outside of any application's private directories into a place where Android 11 facilitates access to after requesting permissions from the user and if the user decides to grant that access.
> 
> I further assume that the system has established some rules on the directory paths which support grant based access for any application that asks for permission. Let's assuem that is `/storage/emulated/0/Public`. What happens if I'd put the maps and themes into `/storage/emulated/0/Public/OfflineMaps` and I grant access to Locus and cgeo, would I then run into the scenario that @MagpieFourtyTwo described, that access to many many small files would take forever?
> 
> I would assume that a directory path with those properties (accessible, after request / grant) to multiple applications, they (G) don't do all the nasty stuff that slows it down so much. If that is true, the whole sync thing wouldn't be necessary, presumably. Please correct me if I'm wrong.

@Lineflyer Where should this topic be discussed further? Depending on the outcome, the topic about creating shadow copies for performance (called "sync" in the given context) may no longer be relevant.

**Comment by eddiemuc** [comment]
My understanding is that we found a valid solution. Is there anything left open ? If yes, we can always open a new issue

**Comment by baiti** [comment]
> My understanding is that we found a valid solution. Is there anything left open ? If yes, we can always open a new issue

Hi @eddiemuc,  If the valid solution involves creating shadow copies for performance, either by default or by presenting the user with the option to do so, then I would be highly motivated to find answers to the questions I posted above. Admittedly they might be more verbose than absolutely necessary.

Let me try as brief as I can: Is the proposed solution required if cgeo is the only app using those maps exclusively?

I would expect the answer is "No" as in this case the data would end up in a directory where cgeo has high performance access to and hence, no shadow copies necessary, correct?

**Comment by Lineflyer** [comment]
> 
> @Lineflyer Where should this topic be discussed further? Depending on the outcome, the topic about creating shadow copies for performance (called "sync" in the given context) may no longer be relevant.

I created #10131 regarding the default paths.
But the topic of performance issue is quite independent of where those files are located as it is a generic problem of scoped storage.

> Hi @eddiemuc, If the valid solution involves creating shadow copies for performance

No, the aligned solution is to **not** sync anything by default, but to keep the sync/buffer as optional feature, e.g. for users who have many unzipped themes and thus experiencing performance issues. Even then, the recommended solution to the user would be to use zipped themes. The sync/buffer option is only a quick fix or a last resort if even zipped is to slow via scoped storage.

**Comment by eddiemuc** [comment]
Yes, I have the same understanding.

I will implement it during next couple of days/weekend.

**Comment by eddiemuc** [comment]
PR #10135 created

**Comment by Lineflyer** [comment]
Thanks @eddiemuc for fixing this in a very good way.
